crates/cranelift/src/compiler/component.rs

```diff
@@ -94,166 +96,230 @@ impl<'a> TrampolineCompiler<'a> {
                 })
             }
             Trampoline::ResourceTransferBorrow => {
-                self.translate_resource_libcall(host::resource_transfer_borrow, |me, rets| {
-                    rets[0] = me.raise_if_resource_trapped(rets[0]);
+                self.translate_host_libcall(host::resource_transfer_borrow, |me, rets| {
+                    rets[0] = me.raise_if_negative_one(rets[0]);
                 })
             }
             Trampoline::ResourceEnterCall => {
-                self.translate_resource_libcall(host::resource_enter_call, |_, _| {})
+                self.translate_host_libcall(host::resource_enter_call, |_, _| {})
             }
             Trampoline::ResourceExitCall => {
-                self.translate_resource_libcall(host::resource_exit_call, |me, rets| {
+                self.translate_host_libcall(host::resource_exit_call, |me, rets| {
                     me.raise_if_host_trapped(rets.pop().unwrap());
                 })
             }
-            Trampoline::AsyncEnterCall => todo!(),
+            Trampoline::SyncEnterCall => self.translate_sync_enter(),
+            Trampoline::SyncExitCall { callback } => self.translate_sync_exit(*callback),
+            Trampoline::AsyncEnterCall => {
+                self.translate_async_enter_or_exit(host::async_enter, None, ir::types::I8)
+            }
             Trampoline::AsyncExitCall {
                 callback,
                 post_return,
-            } => {
-                _ = (callback, post_return);
-                todo!()
-            }
+            } => self.translate_async_enter_or_exit(
+                host::async_exit,
+                Some((*callback, *post_return)),
+                ir::types::I64,
+            ),
             Trampoline::FutureTransfer => {
-                _ = host::future_transfer;
-                todo!()
+                self.translate_host_libcall(host::future_transfer, |me, rets| {
+                    rets[0] = me.raise_if_negative_one(rets[0]);
+                })
             }
             Trampoline::StreamTransfer => {
-                _ = host::stream_transfer;
-                todo!()
+                self.translate_host_libcall(host::stream_transfer, |me, rets| {
+                    rets[0] = me.raise_if_negative_one(rets[0]);
+                })
             }
             Trampoline::ErrorContextTransfer => {
-                _ = host::error_context_transfer;
-                todo!()
+                self.translate_host_libcall(host::error_context_transfer, |me, rets| {
+                    rets[0] = me.raise_if_negative_one(rets[0]);
+                })
             }
         }
     }
 
-    fn translate_lower_import(
-        &mut self,
-        index: LoweredIndex,
-        options: &CanonicalOptions,
-        lower_ty: TypeFuncIndex,
-    ) {
+    fn flat_stream_element_info(&self, ty: TypeStreamTableIndex) -> Option<CanonicalAbiInfo> {
```

Comment:
Mind adding a comment here clarifying that this is just an optimization and it's always valid to return `None` here?